### PR TITLE
update tests

### PR DESCRIPTION
**Pull request by monperrus**
adds contracts, focused assertions, reorder some tests (all CtConsumableFunction tests at the end)

### CODE REVIEW

File: src/test/java/spoon/test/filters/FilterTest.java [hunk 1]
```diff
@@ -613,21 +623,8 @@ public void testElementMapFunction() throws Exception {
...
-		CtClass<?> cls = launcher.getFactory().Class().get(Tacos.class);
-		cls.map((CtClass<?> c, CtConsumer<Object> out)->out.accept(c.getParent()))
-			.forEach((CtElement e)->{
-				assertEquals(cls.getParent(), e);
-			});
-		assertEquals(cls.getParent(), cls.map((CtClass<?> c, CtConsumer<Object> out)->out.accept(c.getParent())).list().get(0));
-	}
-	@Test
 	public void testElementMapFunctionArray() throws Exception {
+		// contract: when a function returns an array, all non-null values are sent to the next step
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/filters/FilterTest.java:627]: I suggest to add another contract here
//contract2: when input is null then mapping function is not called at all.

File: src/test/java/spoon/test/filters/FilterTest.java [hunk 2]
```diff
@@ -613,21 +623,8 @@ public void testElementMapFunction() throws Exception {
...
 		final Launcher launcher = new Launcher();
 		CtQueryImpl q = new CtQueryImpl().map((String s)->new String[]{"a", null, s});
 		List<String> list = q.setInput(null).list();
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/filters/FilterTest.java:630]: or we can move this sub test to own test

File: src/test/java/spoon/test/filters/FilterTest.java [hunk 3]
```diff
@@ -640,82 +637,87 @@ public void testElementMapFunctionArray() throws Exception {
...
 		assertSame(cls, ((CtQueryImpl)q).getInputs().get(0));
-		//change the input of query to cls2
-		q.setInput(cls2).forEach((String name)->{
-			context.count++;
-			assertEquals(cls2.getSimpleName(), name);
-		});
+
+		// now changing the input of query to cls2
+		q.setInput(cls2);
 		//the input is still cls2
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/filters/FilterTest.java:677]: I am evaluating the query twice intentionally. To assure the contract that call of setInput is persistent... that it is not reset after first call of list or forEach. So please add the second call of list again

**Comment by pvojtechovsky** [comment]
Thanks for revision of tests. I will merge it after I get feedback to my comments.

File: src/test/java/spoon/test/filters/FilterTest.java (hunk 3 above)
Comment (monperrus) [reply src/test/java/spoon/test/filters/FilterTest.java:677]: what do you mean, that the input is stored somewhere or that the query output is cached?

File: src/test/java/spoon/test/filters/FilterTest.java (hunk 1 above)
Comment (monperrus) [reply src/test/java/spoon/test/filters/FilterTest.java:627]: interesting, done

File: src/test/java/spoon/test/filters/FilterTest.java (hunk 3 above)
Comment (pvojtechovsky) [reply src/test/java/spoon/test/filters/FilterTest.java:677]: I mean that input is stored

Comment (monperrus) [reply src/test/java/spoon/test/filters/FilterTest.java:677]: I see. done.

Comment (monperrus) [reply src/test/java/spoon/test/filters/FilterTest.java:677]: I see. done

**Pull request merged**